### PR TITLE
Mpnnfreeze

### PR DESCRIPTION
**Pull request by cbilodeau2**
This update is to allow users to freeze mpnn/ffn layers for pretraining. I'm not sure what additional info is needed so let me know if you need anything else for this pull request. Thanks!

### CODE REVIEW

File: chemprop/utils.py [hunk 1]
```diff
@@ -145,6 +146,127 @@ def load_checkpoint(path: str,
...
+    :param loaded_state_dict: state_dict for checkpoint model.
+    :param model_state_dict: state_dict for current model.
+    :param logger: A logger.
+    :return: The updated state_dict for the current model. 
+    """
+    debug = logger.debug if logger is not None else print
+
+    
+    if model_param_name not in model_state_dict:
+        debug(f'Pretrained parameter "{param_name}" cannot be found in model parameters.')
```
Comment (kevingreenman) [inline chemprop/utils.py:167]: I think it should be `model_param_name` instead of `param_name` on line 167

File: chemprop/args.py [hunk 2]
```diff
@@ -265,6 +265,8 @@ class TrainArgs(CommonArgs):
     """Additional metrics to use to evaluate the model. Not used for early stopping."""
     save_dir: str = None
     """Directory where model checkpoints will be saved."""
+    checkpoint_frzn: str = None
+    """Directory where model checkpoints be loaded from for overwriting and freezing weights."""
```
Comment (cjmcgill) [inline chemprop/args.py:269]: Slight language change to specify that it's one checkpoint file and the path is to the file and not the directory

"Path to model checkpoint file to be loaded for overwriting and freezing weights."

File: chemprop/args.py [hunk 3]
```diff
@@ -383,6 +385,22 @@ class TrainArgs(CommonArgs):
...
+    Overwrites weights for the first n layers of the ffn from checkpoint model (specified checkpoint_frzn), 
+    where n is specified in the input.
+    Automatically also freezes mpnn weights. 
+    """
+    freeze_first_only: bool = False
+    """
+    Determines whether or not to use checkpoint_frzn for just the first encoder.
+    Default (False) is to use the checkpoint to freeze all encoders.
+    (only relevant for number_of_molecules > 1, where checkpoint model has number_of_molecules = 1)
+    """
```
Comment (cjmcgill) [inline chemprop/args.py:403]: For the default behavior where it freezes all encoders, this is supposing distinct mpnns for the two molecules or is it one mpnn copied for both?

Comment (cbilodeau2) [reply chemprop/args.py:403]: If the checkpoint file contains two molecules, then the two different MPNNs will be used and frozen. If you are training a model for two molecules, but your frozen checkpoint only includes one molecule, then it will, by default, copy the single MPNN and use it for both molecules and freeze it (unless you select freeze first only, in which case the MPNN will only be used for the first molecule).

File: chemprop/models/model.py [hunk 4]
```diff
@@ -49,7 +49,16 @@ def create_encoder(self, args: TrainArgs) -> None:
         :param args: A :class:`~chemprop.args.TrainArgs` object containing model arguments.
         """
         self.encoder = MPN(args)
-
+              
+        if hasattr(args,'checkpoint_frzn'):
```
Comment (cjmcgill) [inline chemprop/models/model.py:53]: Even if checkpoint_frzn hasn't been provided, it will still have a default value of None and will still trigger this if statement.

File: chemprop/models/model.py [hunk 5]
```diff
@@ -49,7 +49,16 @@ def create_encoder(self, args: TrainArgs) -> None:
...
+            if (args.frzn_mpnn) | (args.frzn_ffn_layers>0):
```
Comment (cjmcgill) [inline chemprop/models/model.py:54]: If you change the first layer of logic check to be a check for None, then you don't need this or statement. You can make this just the greater than zero check.

Stylistically, other places in chemprop where we have "or" operations it's been done with the word "or" instead of "|" so if there was a need to keep the or, then I would recommend changing the style.

File: chemprop/models/model.py [hunk 6]
```diff
@@ -94,9 +103,16 @@ def create_ffn(self, args: TrainArgs) -> None:
                 dropout,
                 nn.Linear(args.ffn_hidden_size, self.output_size),
             ])
+            
 
         # Create FFN model
         self.ffn = nn.Sequential(*ffn)
+        
+        if hasattr(args,'checkpoint_frzn'):
```
Comment (cjmcgill) [inline chemprop/models/model.py:111]: As above, should replace with "if args.checkpoint_frzn is not None"

File: chemprop/models/mpn.py [hunk 7]
```diff
@@ -183,6 +183,7 @@ def __init__(self,
         else:
             self.encoder = nn.ModuleList([MPNEncoder(args, self.atom_fdim, self.bond_fdim)
                                           for _ in range(args.number_of_molecules)])
+
```
Comment (cjmcgill) [inline chemprop/models/mpn.py:186]: I think this added space was made by mistake, would remove it

File: chemprop/train/run_training.py [hunk 8]
```diff
@@ -200,7 +200,12 @@ def run_training(args: TrainArgs,
         else:
             debug(f'Building model {model_idx}')
             model = MoleculeModel(args)
-
+            
+        # Optionally, overwrite weights:
+        if args.frzn_mpnn | (args.frzn_ffn_layers > 0):
```
Comment (cjmcgill) [inline chemprop/train/run_training.py:205]: As above, would recommend replacing "|" with "or"

File: chemprop/models/model.py (hunk 5 above)
Comment (cbilodeau2) [reply chemprop/models/model.py:54]: If the user freezes ffn layers, but not the mpnn, right now the function in utils.py prints a warning that the mpnn will also be frozen and overwrites the mpnn and ffn layers. Because of this if the user freezes ffn layers but not the mpnn, I want this to also trigger freezing the mpnn. Let me know if this doesn't make sense or if I'm wrong (I'm changing the or statements).

File: chemprop/train/run_training.py [hunk 9]
```diff
@@ -200,7 +200,12 @@ def run_training(args: TrainArgs,
...
+            debug(f'Loading mpn parameters from {args.checkpoint_frzn}')
```
Comment (cjmcgill) [inline chemprop/train/run_training.py:206]: I might specify in this log message that it is the frozen parameters. And maybe since it is mpn for ffn parameters, just saying model parameters generally.

File: chemprop/train/run_training.py [hunk 10]
```diff
@@ -200,7 +200,12 @@ def run_training(args: TrainArgs,
...
+            model = load_frzn_model(model=model,path=args.checkpoint_frzn, current_args=args, logger=logger)     
+        
         debug(model)
         debug(f'Number of parameters = {param_count(model):,}')
```
Comment (cjmcgill) [inline chemprop/train/run_training.py:210]: Does this parameter count update acknowledge frozen parameters? I wonder if there's a way we can provide a total parameter count and a frozen parameter count. Not a requirement but would be nice.

File: chemprop/utils.py (hunk 1 above)
Comment (cjmcgill) [inline chemprop/utils.py]: Would we want to raise a hard error at this point if it happened?

Comment (cbilodeau2) [reply chemprop/utils.py]: I was thinking that could be nice, it's pretty easy to miss these errors. I think this is a warning in the load_checkpoint function though, and I was trying to be consistent with respect to what triggers errors/warnings.

**Comment by cjmcgill** [comment]
@cbilodeau2 during your last push it looks like a lot of the changes were replaced with jupyter notebook checkpoints? Can you change back to your normal python files?

File: chemprop/utils.py (hunk 1 above)
Comment (cjmcgill) [reply chemprop/utils.py]: Very good then. I hadn't yet gotten to the load_checkpoint file. Having them all in one place and consistent makes a ton of sense.

File: chemprop/args.py (hunk 3 above)
Comment (cjmcgill) [reply chemprop/args.py:403]: That's a great way to handle that

File: chemprop/train/run_training.py [hunk 11]
```diff
@@ -200,7 +200,12 @@ def run_training(args: TrainArgs,
         else:
             debug(f'Building model {model_idx}')
             model = MoleculeModel(args)
-
+            
+        # Optionally, overwrite weights:
+        if args.frzn_mpnn or (args.frzn_ffn_layers > 0):
```
Comment (cjmcgill) [inline chemprop/train/run_training.py:205]: This doesn't include a check for if there's a checkpoint_frzn path. If that's not provided you'll hit an error when you try to load the checkpoint but it won't be an informative error. I think you need to include that path in this check or raise an informative error somewhere.

File: chemprop/args.py [hunk 12]
```diff
@@ -383,6 +385,22 @@ class TrainArgs(CommonArgs):
     """Overwrites the default atom descriptors with the new ones instead of concatenating them"""
     no_bond_features_scaling: bool = False
     """Turn off atom feature scaling."""
+    frzn_mpnn: bool = False
```
Comment (cjmcgill) [inline chemprop/args.py:388]: Can we collapse this argument into other arguments?

As set up now, mpnn is frozen any time anything is frozen. And we need a checkpoint_frzn any time anything is frozen. So we could just key off of whether there is a checkpoint_frzn to decide if we are freezing the mpnn.

If we had situations where mpnn was not frozen but ffn was, then you would need this extra degree of freedom. But we don't have it currently configured in a way right now where it's possible since the frzn_ffn will override a false on frzn_mpnn. I don't think that's an option that we actually need in the future, so I think we're safe to remove this argument.

File: chemprop/train/run_training.py (hunk 10 above)
Comment (cbilodeau2) [reply chemprop/train/run_training.py]: Added this!

File: chemprop/train/run_training.py [hunk 13]
```diff
@@ -200,9 +200,15 @@ def run_training(args: TrainArgs,
...
+            
+        # Optionally, overwrite weights:
+        if args.checkpoint_frzn is not None:
+            debug(f'Loading and freezing parameters from {args.checkpoint_frzn}.')
+            model = load_frzn_model(model=model,path=args.checkpoint_frzn, current_args=args, logger=logger)     
+        
         debug(model)
-        debug(f'Number of parameters = {param_count(model):,}')
+        debug(f'Number of unfrozen parameters = {param_count(model):,}')
+        debug(f'Total number of parameters = {param_count_all(model):,}')
```
Comment (cjmcgill) [inline chemprop/train/run_training.py:211]: This is a nice change to these functions.

Most users will not be using the frozen functionality. Might be nice to make it so that the output reads without reference to frozen if you're not using frozen. Maybe have it inside an if/else statement.

File: chemprop/utils.py [hunk 14]
```diff
@@ -145,6 +146,122 @@ def load_checkpoint(path: str,
...
+    model_state_dict = model.state_dict()
+    
+    if loaded_args.number_of_molecules==1 & current_args.number_of_molecules==1:      
+        encoder_param_names = ['encoder.encoder.0.W_i.weight', 'encoder.encoder.0.W_h.weight', 'encoder.encoder.0.W_o.weight', 'encoder.encoder.0.W_o.bias']
+        if current_args.checkpoint_frzn is not None:
+            # Freeze the MPNN
+            for param_name in encoder_param_names:
+                model_state_dict = overwrite_state_dict(param_name,param_name,loaded_state_dict,model_state_dict)
+            
+        if current_args.frzn_ffn_layers > 0:
```
Comment (cjmcgill) [inline chemprop/utils.py:208]: I ran some tests for what happens when you specify a larger number of frozen ffn than you have layers. If you set number of layers >= your number of layers, you get an error when it tries to do gradient descent but nothing has saved gradients. Maybe raise an error somewhere in here that has a descriptive message about the number of ffn layers

**Review by cjmcgill — CHANGES_REQUESTED** [review]
Hi @cbilodeau2 took me a little while to put together a good test and take a close look at everything. Thanks this is looking really good and it tested successfully for me. Great job.

I made two notes for additional changes that might be nice to add.

**Comment by cbilodeau2** [comment]
> 
> 
> Hi @cbilodeau2 took me a little while to put together a good test and take a close look at everything. Thanks this is looking really good and it tested successfully for me. Great job.
> 
> I made two notes for additional changes that might be nice to add.

Thanks!! I just added in those two edits, let me know if you have any other suggestions.

**Review by cjmcgill — APPROVED** [review]
Looks good!

**Comment by cjmcgill** [comment]
@cbilodeau2 Okay now a couple of last steps to clean up things before I merge. (Possibly I should have said this before I marked it as accepted). These are things to try and keep the merge history for the main branch easy to read and follow. These things are easier to do if you use a git gui like gitkraken. If you need help doing these let me know (good experience for you to give it a try yourself).

1. Rebase your branch onto the most recent version of the master branch.
2. Squash your commits in your branch together into a single commit with a descriptive name.
3. Force push your rebased and squashed commit to the origin (the remote) to overwrite the version of the branch with multiple commits.

**Comment by cjmcgill** [comment]
I have not done this by command line but @hesther had advice for how to do it for another recent PR. I'm quoting it below. This was advice for how to squash a 3-commit branch. Yours is 10 commits so you would use 10 in place of 3.

> Looks great, thanks! Could you squash the three commits into one?
> 
> If you are not familiar with doing so:
> `git rebase -i HEAD~3`
> this will open an editor and show the three latest commits (starting with the newest one). Put the second and third on "s" instead of "pick". Save and close. This will open up a new editor which shows the three latest commit messages. Since the changes were all super small and bug fixes I suggest you comment them out apart from the first commit message (Fix unknown keyword argument passed to accuracy in train.evaluate). Save and close. Use
> `git log`
> to check whether the commit history is what you want it to be (a single commit instead of three individual commits). If it is ok, push your changes. Since your repo has now diverged from the upstream repo, you need to force push:
> `git push -f`

**Comment by hesther** [comment]
Just wanted to say that this quote only covers steps 2 and 3, so you have to first merge in the master branch (step 1)
